deploy/init-db/db.sql.txt

```diff
@@ -0,0 +1,62 @@
+-- 'resources' is used to store all of our resources. See comments below for an explanation of the columns.
+CREATE TABLE resources (
+    -- example: "/planes/radius/local/resourceGroups/rg1/providers/Applications.Core/applications/my-app"
+    --
+    -- We use columns to break out the most important components of the resource id for optimal querying.
+    --
+    -- Since resource ids are case-insensitive we canonicalize these columns to lowercase.
+    -- We store the original resource id with the original casing so users can work with their preferred
+    -- naming/casing conventions.
+    --
+    -- We ensure a leading and trailing slash on the components of the resource id for ease of comparison.
+    --
+    -- id -> "/planes/radius/local/resourcegroups/rg1/providers/applications.core/applications/my-app/"
+    -- resource_type -> "/applications.core/applications/"
+    -- root_scope -> "/planes/radius/local/resourcegroups/rg1/"
+    -- routing_scope -> "/applications.core/applications/my-app/"
+
+    -- resource id used as key.
+    id TEXT PRIMARY KEY NOT NULL,
```

Comment:
Can we discuss using the resource_id as a primary key in PostGres versus say a UUID while creating a separate unique index for the resource id field??
Concerns are length of id and string type for primary key where size will increase and performance will be degraded over time

Reply:
Sounds like a good topic. 

I started with the resource id as the PK because it supports all of our access patterns. Some of the prototyping that I did actually added a UUID, which would be suitable as a PK.

The UUID is helpful because it allows you to tell the difference between different instances of a resource with the same name (over time). [Kubernetes does this](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/).  

Take a look a the Go APIs of the datastore - they map 1:1 with the operations exposed by the RP.

- Resource List (some scope) -> Query resources by type + scope
- Resource Get (some id) -> Get resource by id
- Resource Put (some id) -> Save resource by id
- Resource Delete (some id) -> Delete resource by id

If we used in a synthetic key like a UUID we'd have to reckon with how we map from resource id -> UUID. We'd have a UUID per-resource but it would almost never be used for SQL statements. We'd still need an index on the id and would be primarily leveraging that.

Any ideas?

Reply:
Keep in mind we're not going to be using RDBMS features like joins. We're treating PostgreSQL like a document store.

Reply:
Do we have an idea on what the maximum table size will be?

Reply:
I'd imagine a 100k or so would be a **really** **really** large amount of data for a single deployment of Radius. You can think of one row = one resource in Radius. 

Postgres is one of a few options we'll support. You can imagine another use case where the user chooses CosmosDB or DynamoDB or something else that has sharding built-in natively. There are good natural ways to use resource ids for sharding.

---

Totally up for debate .. some additional context from me.

- I wanted to avoid doing DDL dynamically. This would be required if we sharded on resource type.
- I think Postgres is a good **default** choice for Radius. Users that are really invested might choose something that can scale better.

Reply:
The table size being limited, resource id as a primary key is a reasonable approach.  
Question on 'Adding in a UUID in the future', if resource id is a primary key, we will always have only 1 value of UUID in the table at any given time. Unless we further build out and store history/audit of data updates, I am not seeing where/how it will be useful. Is that a possible future usecase ?

Reply:
The reason I linked the Kuberentes `uid` feature is to give you can idea of the usecase. 

It allows the control-plane to know the difference between difference instances of the resource that are created at different times. Example:

- A user creates resource `A`, it gets uuid: `12E67073-1BE4-4935-80AC-6C7D02BE5D8A`
- A user deletes resource `A`
- A user creates resource `A`, it gets uuid: `6FE3910B-3185-43EC-82C3-70902FEF3421`

It allows for a form of opportunistic concurrency control that's based on resource identity rather than resource state.

Reply:
So it sounds like we've largely bottomed-out on this topic... with the consensus that we'll use the resource id as the PK. 

Does that sound right? Anything else we should consider?